driver-core/src/main/java/com/datastax/driver/core/SessionManager.java

```diff
@@ -294,34 +291,39 @@ private boolean replacePool(Host host, HostDistance distance, HostConnectionPool
 
-        // Creating a pool is somewhat long since it has to create the connection, so do it asynchronously.
-        return executor.submit(new Callable<Boolean>() {
-            public Boolean call() {
-                try {
-                    while (true) {
-                        HostConnectionPool previous = pools.get(host);
-                        if (previous != null && !previous.isClosed())
-                            return true;
-
-                        if (replacePool(host, distance, previous)) {
-                            logger.debug("Added connection pool for {}", host);
-                            return true;
+        while (true) {
+            previous = pools.get(host);
+            if (previous != null && !previous.isClosed())
+                return Futures.immediateFuture(true);
+
+            final SettableFuture<Boolean> future = SettableFuture.create();
+            ListenableFuture<Void> newPoolInit = replacePool(host, distance, previous);
+            if (newPoolInit != null) {
+                Futures.addCallback(newPoolInit, new FutureCallback<Void>() {
+                    @Override
+                    public void onSuccess(Void result) {
+                        logger.debug("Added connection pool for {}", host);
+                        future.set(true);
+                    }
+
+                    @Override
+                    public void onFailure(Throwable t) {
+                        if (t instanceof UnsupportedProtocolVersionException) {
+                            cluster.manager.logUnsupportedVersionProtocol(host);
+                            cluster.manager.triggerOnDown(host, false);
+                        } else if (t instanceof ClusterNameMismatchException) {
+                            ClusterNameMismatchException e = (ClusterNameMismatchException)t;
+                            cluster.manager.logClusterNameMismatch(host, e.expectedClusterName, e.actualClusterName);
+                            cluster.manager.triggerOnDown(host, false);
+                        } else {
+                            logger.error("Error creating pool to " + host, t);
                         }
+                        future.set(false);
                     }
-                } catch (UnsupportedProtocolVersionException e) {
-                    cluster.manager.logUnsupportedVersionProtocol(host);
-                    cluster.manager.triggerOnDown(host, false);
-                    return false;
-                } catch (ClusterNameMismatchException e) {
-                    cluster.manager.logClusterNameMismatch(host, e.expectedClusterName, e.actualClusterName);
-                    cluster.manager.triggerOnDown(host, false);
-                    return false;
-                } catch (Exception e) {
-                    logger.error("Error creating pool to " + host, e);
-                    return false;
-                }
+                }, executor);
```

Comment:
Here, contrary to forceRenewPool, we are using the blocking executor - is this per design? The task doesn't seem to require it.